Broken tests in Mac with apple-clang 10.0 as default [PR 4332] [BROKEN DEVELOP]

⚠️ ⚠️  PR #4332 breaks in my Mac, there is a test that fails (commit hash e2246c90):

> CMakeGeneratorTest::test_cmake_default_generator_osx

The error is the automatic detection of the clang version by CMake:

```bash
CMake Error at conanbuildinfo.cmake:300 (message):
  Incorrect 'apple-clang' version 'compiler.version=9.0' is not the one
  detected by CMake: 'Clang=10.0'
```

and the full log:

```bash

Error
Traceback (most recent call last):
  File "/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/lib/python3.7/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/lib/python3.7/unittest/case.py", line 615, in run
    testMethod()
  File "/Users/jgsogo/dev/conan/conan/conans/test/functional/generators/cmake_generator_test.py", line 72, in test_cmake_default_generator_osx
    self._check_build_generator("Macos", "apple-clang", "9.0", "Unix Makefiles")
  File "/Users/jgsogo/dev/conan/conan/conans/test/functional/generators/cmake_generator_test.py", line 58, in _check_build_generator
    client.run("build .")
  File "/Users/jgsogo/dev/conan/conan/conans/test/utils/tools.py", line 615, in run
    raise Exception(exc_message)
Exception: 
------------------------ Command failed (unexpectedly): ------------------------
build .
----------------------------------- Output: ------------------------------------
conanfile.py (foo/0.1@None/None): Running build()

----Running------
> cd '/private/var/folders/yq/14hmvxm96xd7gfgl37_tnrbh0000gn/T/tmpsmfbdzzoconans/path with spaces' && cmake -G "Unix Makefiles" -DCONAN_EXPORTED="1" -DCONAN_IN_LOCAL_CACHE="OFF" -DCONAN_COMPILER="apple-clang" -DCONAN_COMPILER_VERSION="9.0" -DCONAN_LIBCXX="libstdc++" -DCMAKE_INSTALL_PREFIX="/private/var/folders/yq/14hmvxm96xd7gfgl37_tnrbh0000gn/T/tmpsmfbdzzoconans/path with spaces/package" -DCMAKE_INSTALL_BINDIR="bin" -DCMAKE_INSTALL_SBINDIR="bin" -DCMAKE_INSTALL_LIBEXECDIR="bin" -DCMAKE_INSTALL_LIBDIR="lib" -DCMAKE_INSTALL_INCLUDEDIR="include" -DCMAKE_INSTALL_OLDINCLUDEDIR="include" -DCMAKE_INSTALL_DATAROOTDIR="share" -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY="ON" -Wno-dev '/private/var/folders/yq/14hmvxm96xd7gfgl37_tnrbh0000gn/T/tmpsmfbdzzoconans/path with spaces'
-----------------
Logging command output to file '/private/var/folders/yq/14hmvxm96xd7gfgl37_tnrbh0000gn/T/tmpsmfbdzzoconans/path with spaces/conan_run.log'
-- The CXX compiler identification is AppleClang 10.0.0.10001145
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Conan: called by CMake conan helper
CMake Error at conanbuildinfo.cmake:300 (message):
  Incorrect 'apple-clang' version 'compiler.version=9.0' is not the one
  detected by CMake: 'Clang=10.0'
Call Stack (most recent call first):
  conanbuildinfo.cmake:370 (conan_error_compiler_version)
  conanbuildinfo.cmake:441 (check_compiler_version)
  conanbuildinfo.cmake:59 (conan_check_compiler)
  CMakeLists.txt:6 (conan_basic_setup)


-- Configuring incomplete, errors occurred!
See also "/private/var/folders/yq/14hmvxm96xd7gfgl37_tnrbh0000gn/T/tmpsmfbdzzoconans/path with spaces/CMakeFiles/CMakeOutput.log".
ERROR: conanfile.py (foo/0.1@None/None): Error in build() method, line 12
	cmake.configure()
	ConanException: Error 1 while executing cd '/private/var/folders/yq/14hmvxm96xd7gfgl37_tnrbh0000gn/T/tmpsmfbdzzoconans/path with spaces' && cmake -G "Unix Makefiles" -DCONAN_EXPORTED="1" -DCONAN_IN_LOCAL_CACHE="OFF" -DCONAN_COMPILER="apple-clang" -DCONAN_COMPILER_VERSION="9.0" -DCONAN_LIBCXX="libstdc++" -DCMAKE_INSTALL_PREFIX="/private/var/folders/yq/14hmvxm96xd7gfgl37_tnrbh0000gn/T/tmpsmfbdzzoconans/path with spaces/package" -DCMAKE_INSTALL_BINDIR="bin" -DCMAKE_INSTALL_SBINDIR="bin" -DCMAKE_INSTALL_LIBEXECDIR="bin" -DCMAKE_INSTALL_LIBDIR="lib" -DCMAKE_INSTALL_INCLUDEDIR="include" -DCMAKE_INSTALL_OLDINCLUDEDIR="include" -DCMAKE_INSTALL_DATAROOTDIR="share" -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY="ON" -Wno-dev '/private/var/folders/yq/14hmvxm96xd7gfgl37_tnrbh0000gn/T/tmpsmfbdzzoconans/path with spaces'

--------------------------------------------------------------------------------


E
======================================================================
ERROR: test_cmake_default_generator_osx (conan.conans.test.functional.generators.cmake_generator_test.CMakeGeneratorTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jgsogo/dev/conan/conan/conans/test/functional/generators/cmake_generator_test.py", line 72, in test_cmake_default_generator_osx
    self._check_build_generator("Macos", "apple-clang", "9.0", "Unix Makefiles")
  File "/Users/jgsogo/dev/conan/conan/conans/test/functional/generators/cmake_generator_test.py", line 58, in _check_build_generator
    client.run("build .")
  File "/Users/jgsogo/dev/conan/conan/conans/test/utils/tools.py", line 615, in run
    raise Exception(exc_message)
Exception: 
------------------------ Command failed (unexpectedly): ------------------------
build .
----------------------------------- Output: ------------------------------------
conanfile.py (foo/0.1@None/None): Running build()

----Running------
> cd '/private/var/folders/yq/14hmvxm96xd7gfgl37_tnrbh0000gn/T/tmpsmfbdzzoconans/path with spaces' && cmake -G "Unix Makefiles" -DCONAN_EXPORTED="1" -DCONAN_IN_LOCAL_CACHE="OFF" -DCONAN_COMPILER="apple-clang" -DCONAN_COMPILER_VERSION="9.0" -DCONAN_LIBCXX="libstdc++" -DCMAKE_INSTALL_PREFIX="/private/var/folders/yq/14hmvxm96xd7gfgl37_tnrbh0000gn/T/tmpsmfbdzzoconans/path with spaces/package" -DCMAKE_INSTALL_BINDIR="bin" -DCMAKE_INSTALL_SBINDIR="bin" -DCMAKE_INSTALL_LIBEXECDIR="bin" -DCMAKE_INSTALL_LIBDIR="lib" -DCMAKE_INSTALL_INCLUDEDIR="include" -DCMAKE_INSTALL_OLDINCLUDEDIR="include" -DCMAKE_INSTALL_DATAROOTDIR="share" -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY="ON" -Wno-dev '/private/var/folders/yq/14hmvxm96xd7gfgl37_tnrbh0000gn/T/tmpsmfbdzzoconans/path with spaces'
-----------------
Logging command output to file '/private/var/folders/yq/14hmvxm96xd7gfgl37_tnrbh0000gn/T/tmpsmfbdzzoconans/path with spaces/conan_run.log'
-- The CXX compiler identification is AppleClang 10.0.0.10001145
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Conan: called by CMake conan helper
CMake Error at conanbuildinfo.cmake:300 (message):
  Incorrect 'apple-clang' version 'compiler.version=9.0' is not the one
  detected by CMake: 'Clang=10.0'
Call Stack (most recent call first):
  conanbuildinfo.cmake:370 (conan_error_compiler_version)
  conanbuildinfo.cmake:441 (check_compiler_version)
  conanbuildinfo.cmake:59 (conan_check_compiler)
  CMakeLists.txt:6 (conan_basic_setup)


-- Configuring incomplete, errors occurred!
See also "/private/var/folders/yq/14hmvxm96xd7gfgl37_tnrbh0000gn/T/tmpsmfbdzzoconans/path with spaces/CMakeFiles/CMakeOutput.log".
ERROR: conanfile.py (foo/0.1@None/None): Error in build() method, line 12
	cmake.configure()
	ConanException: Error 1 while executing cd '/private/var/folders/yq/14hmvxm96xd7gfgl37_tnrbh0000gn/T/tmpsmfbdzzoconans/path with spaces' && cmake -G "Unix Makefiles" -DCONAN_EXPORTED="1" -DCONAN_IN_LOCAL_CACHE="OFF" -DCONAN_COMPILER="apple-clang" -DCONAN_COMPILER_VERSION="9.0" -DCONAN_LIBCXX="libstdc++" -DCMAKE_INSTALL_PREFIX="/private/var/folders/yq/14hmvxm96xd7gfgl37_tnrbh0000gn/T/tmpsmfbdzzoconans/path with spaces/package" -DCMAKE_INSTALL_BINDIR="bin" -DCMAKE_INSTALL_SBINDIR="bin" -DCMAKE_INSTALL_LIBEXECDIR="bin" -DCMAKE_INSTALL_LIBDIR="lib" -DCMAKE_INSTALL_INCLUDEDIR="include" -DCMAKE_INSTALL_OLDINCLUDEDIR="include" -DCMAKE_INSTALL_DATAROOTDIR="share" -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY="ON" -Wno-dev '/private/var/folders/yq/14hmvxm96xd7gfgl37_tnrbh0000gn/T/tmpsmfbdzzoconans/path with spaces'

--------------------------------------------------------------------------------


----------------------------------------------------------------------
Ran 1 test in 1.141s

FAILED (errors=1)

Process finished with exit code 1

```

--- a/conans/test/functional/generators/cmake_generator_test.py
+++ b/conans/test/functional/generators/cmake_generator_test.py
@@ -35,28 +35,37 @@ conan_basic_setup()
 add_executable(dummy dummy.cpp)
 """
 
+PROFILE = """
+[settings]
+os_build={os_build}
+compiler={compiler}
+compiler.version={compiler_version}
+compiler.libcxx=libstdc++
+"""
+
 
 class CMakeGeneratorTest(unittest.TestCase):
 
-    def _check_build_generator(self, generator):
+    def _check_build_generator(self, os_build, compiler, compiler_version, generator):
         client = TestClient()
         client.save({"conanfile.py": CONAN_RECIPE,
                      "CMakeLists.txt": CMAKE_RECIPE,
-                     "dummy.cpp": CPP_CONTENT})
-
-        client.run("install .")
+                     "dummy.cpp": CPP_CONTENT,
+                     "my_profile": PROFILE.format(os_build=os_build, compiler=compiler,
+                                                  compiler_version=compiler_version)})
+        client.run("install . -p my_profile")
         client.run("build .")
         self.assertIn("Check for working CXX compiler", client.out)
         self.assertIn('cmake -G "{}"'.format(generator), client.out)
 
     @unittest.skipUnless(tools.os_info.is_linux, "Compilation with real gcc needed")
     def test_cmake_default_generator_linux(self):
-        self._check_build_generator("Unix Makefiles")
+        self._check_build_generator("Linux", "gcc", "5", "Unix Makefiles")
 
     @unittest.skipUnless(tools.os_info.is_windows, "MinGW is only supported on Windows")
     def test_cmake_default_generator_windows(self):
-        self._check_build_generator("MinGW Makefiles")
+        self._check_build_generator("Windows", "gcc", "7", "MinGW Makefiles")
 
     @unittest.skipUnless(tools.os_info.is_macos, "Compilation with real clang is needed")
     def test_cmake_default_generator_osx(self):
-        self._check_build_generator("Unix Makefiles")
+        self._check_build_generator("Macos", "apple-clang", "9.0", "Unix Makefiles")

--- a/conans/test/functional/generators/cmake_generator_test.py
+++ b/conans/test/functional/generators/cmake_generator_test.py
@@ -16,7 +16,7 @@ class FooConan(ConanFile):
     generators = "cmake"
 
     def build(self):
-        cmake = CMake(self, generator="{}")
+        cmake = CMake(self)
         cmake.configure()
 """
 
@@ -40,7 +40,7 @@ class CMakeGeneratorTest(unittest.TestCase):
 
     def _check_build_generator(self, generator):
         client = TestClient()
-        client.save({"conanfile.py": CONAN_RECIPE.format(generator),
+        client.save({"conanfile.py": CONAN_RECIPE,
                      "CMakeLists.txt": CMAKE_RECIPE,
                      "dummy.cpp": CPP_CONTENT})
 

--- a/conans/test/functional/generators/cmake_generator_test.py
+++ b/conans/test/functional/generators/cmake_generator_test.py
@@ -1,4 +1,5 @@
 import unittest
+import os
 
 from nose.plugins.attrib import attr
 
@@ -12,7 +13,7 @@ from conans import ConanFile, CMake
 class FooConan(ConanFile):
     name = "foo"
     version = "0.1"
-    settings = "os_build", "compiler", "build_type"
+    settings = "os_build"
     generators = "cmake"
 
     def build(self):
@@ -27,45 +28,41 @@ int main() {}
 
 CMAKE_RECIPE = """
 cmake_minimum_required(VERSION 2.8.12)
-project(dummy CXX)
+project(dummy)
 
 include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
 conan_basic_setup()
-
-add_executable(dummy dummy.cpp)
 """
 
 PROFILE = """
 [settings]
+os={os_build}
 os_build={os_build}
-compiler={compiler}
-compiler.version={compiler_version}
-compiler.libcxx=libstdc++
 """
-
 
 class CMakeGeneratorTest(unittest.TestCase):
 
-    def _check_build_generator(self, os_build, compiler, compiler_version, generator):
+    def _check_build_generator(self, os_build, generator):
         client = TestClient()
         client.save({"conanfile.py": CONAN_RECIPE,
                      "CMakeLists.txt": CMAKE_RECIPE,
                      "dummy.cpp": CPP_CONTENT,
-                     "my_profile": PROFILE.format(os_build=os_build, compiler=compiler,
-                                                  compiler_version=compiler_version)})
+                     "my_profile": PROFILE.format(os_build=os_build)
+                     })
         client.run("install . -p my_profile")
         client.run("build .")
-        self.assertIn("Check for working CXX compiler", client.out)
+
         self.assertIn('cmake -G "{}"'.format(generator), client.out)
+        self.assertTrue(os.path.isfile(os.path.join(client.current_folder, "Makefile")))
 
     @unittest.skipUnless(tools.os_info.is_linux, "Compilation with real gcc needed")
     def test_cmake_default_generator_linux(self):
-        self._check_build_generator("Linux", "gcc", "5", "Unix Makefiles")
+        self._check_build_generator("Linux", "Unix Makefiles")
 
     @unittest.skipUnless(tools.os_info.is_windows, "MinGW is only supported on Windows")
     def test_cmake_default_generator_windows(self):
-        self._check_build_generator("Windows", "gcc", "7", "MinGW Makefiles")
+        self._check_build_generator("Windows", "MinGW Makefiles")
 
     @unittest.skipUnless(tools.os_info.is_macos, "Compilation with real clang is needed")
     def test_cmake_default_generator_osx(self):
-        self._check_build_generator("Macos", "apple-clang", "9.0", "Unix Makefiles")
+        self._check_build_generator("Macos", "Unix Makefiles")

--- a/conans/test/functional/generators/cmake_generator_test.py
+++ b/conans/test/functional/generators/cmake_generator_test.py
@@ -35,37 +35,28 @@ conan_basic_setup()
 add_executable(dummy dummy.cpp)
 """
 
-PROFILE = """
-[settings]
-os_build={os_build}
-compiler={compiler}
-compiler.version={compiler_version}
-compiler.libcxx=libstdc++
-"""
-
 
 class CMakeGeneratorTest(unittest.TestCase):
 
-    def _check_build_generator(self, os_build, compiler, compiler_version, generator):
+    def _check_build_generator(self, generator):
         client = TestClient()
         client.save({"conanfile.py": CONAN_RECIPE,
                      "CMakeLists.txt": CMAKE_RECIPE,
-                     "dummy.cpp": CPP_CONTENT,
-                     "my_profile": PROFILE.format(os_build=os_build, compiler=compiler,
-                                                  compiler_version=compiler_version)})
-        client.run("install . -p my_profile")
+                     "dummy.cpp": CPP_CONTENT})
+
+        client.run("install .")
         client.run("build .")
         self.assertIn("Check for working CXX compiler", client.out)
         self.assertIn('cmake -G "{}"'.format(generator), client.out)
 
     @unittest.skipUnless(tools.os_info.is_linux, "Compilation with real gcc needed")
     def test_cmake_default_generator_linux(self):
-        self._check_build_generator("Linux", "gcc", "5", "Unix Makefiles")
+        self._check_build_generator("Unix Makefiles")
 
     @unittest.skipUnless(tools.os_info.is_windows, "MinGW is only supported on Windows")
     def test_cmake_default_generator_windows(self):
-        self._check_build_generator("Windows", "gcc", "7", "MinGW Makefiles")
+        self._check_build_generator("MinGW Makefiles")
 
     @unittest.skipUnless(tools.os_info.is_macos, "Compilation with real clang is needed")
     def test_cmake_default_generator_osx(self):
-        self._check_build_generator("Macos", "apple-clang", "9.0", "Unix Makefiles")
+        self._check_build_generator("Unix Makefiles")

--- a/conans/test/functional/generators/cmake_generator_test.py
+++ b/conans/test/functional/generators/cmake_generator_test.py
@@ -16,7 +16,7 @@ class FooConan(ConanFile):
     generators = "cmake"
 
     def build(self):
-        cmake = CMake(self)
+        cmake = CMake(self, generator="{}")
         cmake.configure()
 """
 
@@ -40,7 +40,7 @@ class CMakeGeneratorTest(unittest.TestCase):
 
     def _check_build_generator(self, generator):
         client = TestClient()
-        client.save({"conanfile.py": CONAN_RECIPE,
+        client.save({"conanfile.py": CONAN_RECIPE.format(generator),
                      "CMakeLists.txt": CMAKE_RECIPE,
                      "dummy.cpp": CPP_CONTENT})
 

--- a/conans/test/unittests/client/build/cmake_flags.py
+++ b/conans/test/unittests/client/build/cmake_flags.py
@@ -1,7 +1,8 @@
 import unittest
 
+from conans import CMake
 from conans.client.build.cmake_flags import get_generator
-from conans.test.utils.conanfile import MockSettings
+from conans.test.utils.conanfile import MockSettings, ConanFileMock
 
 
 class CMakeGeneratorTest(unittest.TestCase):
@@ -10,13 +11,24 @@ class CMakeGeneratorTest(unittest.TestCase):
         settings = MockSettings({"os_build": "Linux"})
         generator = get_generator(settings)
         self.assertEquals("Unix Makefiles", generator)
+        self._test_default_generator_cmake(settings, "Unix Makefiles")
 
     def test_cmake_default_generator_osx(self):
         settings = MockSettings({"os_build": "Macos"})
         generator = get_generator(settings)
         self.assertEquals("Unix Makefiles", generator)
+        self._test_default_generator_cmake(settings, "Unix Makefiles")
 
     def test_default_generator_windows(self):
         settings = MockSettings({"os_build": "Windows"})
         generator = get_generator(settings)
         self.assertEquals("MinGW Makefiles", generator)
+        self._test_default_generator_cmake(settings, "MinGW Makefiles")
+
+    def _test_default_generator_cmake(self, settings, generator):
+        conanfile = ConanFileMock()
+        conanfile.settings = settings
+        cmake = CMake(conanfile)
+        self.assertEquals(generator, cmake.generator)
+        self.assertIn('-G "{}"'.format(generator), cmake.command_line)
+

